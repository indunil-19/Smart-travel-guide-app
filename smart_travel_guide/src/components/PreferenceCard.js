import React, { useState, useEffect } from "react";
import { View, StyleSheet, Alert } from "react-native";
import {
  Button,
  Card,
  Paragraph,
  Checkbox,
  IconButton,
} from "react-native-paper";
import { theme } from "../core/theme";
import { Picker } from "@react-native-picker/picker";
import { LinearGradient } from "expo-linear-gradient";
import { icon } from "../components/Icon";

const showAlert = () =>
  Alert.alert(
    "Preference not selected",
    "Select at least one preference.",
    [
      {
        text: "Cancel",
        style: "cancel",
      },
    ],
    {
      cancelable: true,
    }
  );

const PreferenceCard = (props) => {
  const [buttonStyle, setButtonStyle] = useState("space-between");
  const [selectedValue, setSelectedValue] = useState();
  const [checkedState, setCheckedState] = useState(
    new Array(props.checkerOptions.length).fill(false)
  );

  useEffect(() => {
    if (props.firstPage && !props.lastPage) {
      setButtonStyle("flex-end");
    } else {
      setButtonStyle("space-between");
    }
  });
  const renderIcon = (option) => {
    return <IconButton icon={icon(option)} color="green" size={22} disabled />;
  };

  const checkHandler = (key, restriction) => {
    if (
      checkedState.reduce(function (acc, val) {
        return acc + val;
      }, 0) >= restriction
    ) {
      if (checkedState[key]) {
        const updatedCheckedState = checkedState.map((item, index) =>
          index === key ? !item : item
        );
        setCheckedState(updatedCheckedState);
      }
    } else {
      const updatedCheckedState = checkedState.map((item, index) =>
        index === key ? !item : item
      );
      setCheckedState(updatedCheckedState);
    }
  };

  const renderButtons = () => {
    // * Renders Buttons on the card according to the type of page
    if (!props.firstPage && !props.lastPage) {
      return (
        <>
          <Button
            icon="skip-previous"
            onPress={() => {
              onButtonPressHandler("prev");
            }}
          >
            Prev
          </Button>
          <Button
            icon="skip-next"
            contentStyle={{ flexDirection: "row-reverse" }}
            onPress={() => {
              onButtonPressHandler("next");
            }}
          >
            Next
          </Button>
        </>
      );
    } else if (props.firstPage && !props.lastPage) {
      return (
        <Button
          icon="skip-next"
          contentStyle={{ flexDirection: "row-reverse" }}
          onPress={() => {
            onButtonPressHandler("next");
          }}
        >
          Next
        </Button>
      );
    } else if (!props.firstPage && props.lastPage) {
      return (
        <>
          <Button
            icon="skip-previous"
            onPress={() => {
              onButtonPressHandler("prev");
            }}
          >
            Prev
          </Button>
          <Button
            icon="checkbox-marked-circle-outline"
            contentStyle={{ flexDirection: "row-reverse" }}
            onPress={() => {
              onButtonPressHandler("done");
            }}
          >
            Done
          </Button>
        </>
      );
    }
  };

  const renderOptions = () => {
    //* Renders a checkbox or dropdown picker according to the question type
    if (props.checkerOptions.length) {
      return props.checkerOptions.map((option, index) => (
        <View key={index} style={styles.optionContainer}>
          <Checkbox.Item
            status={checkedState[index] ? "checked" : "unchecked"}
            label={option}
            color={theme.colors.primary}
            position="leading"
            uncheckedColor={theme.colors.secondary}
            onPress={() => {
              checkHandler(index, props.restriction);
            }}
          />
          {renderIcon(option)}
        </View>
      ));
    } else {
      return (
        <Picker
          dropdownIconColor="#000000"
          selectedValue={selectedValue}
          onValueChange={(itemValue) => setSelectedValue(itemValue)}
        >
          {props.pickerOptions.map((option, index) => (
            <Picker.Item key={index} label={option} value={option} />
          ))}
        </Picker>
      );
    }
  };
  const filterCheckOptions = () => {
    let selectedOptions = [];
    for (let i = 0; i < checkedState.length; i++) {
      if (checkedState[i]) {
        selectedOptions.push(props.checkerOptions[i]);
      }
    }
    return selectedOptions;
  };
  const onButtonPressHandler = (navigate) => {
    if (navigate === "prev") {
      props.navigation("prev");
    } else if (navigate === "next" || navigate === "done") {
      if (!selectedValue && checkedState.some((state) => state)) {
        props.navigation(filterCheckOptions());
      } else if (selectedValue && !checkedState.some((state) => state)) {
        props.navigation(selectedValue);
      } else {
        showAlert();
      }
    }
  };

  return (
    <Card mode="outlined" style={{ ...styles.card, display: props.display }}>
      <LinearGradient
        colors={["#8af4e5", "white"]}
        style={styles.linearGradient}
      >
        <Card.Content>
          <Paragraph style={styles.question}>{props.question}</Paragraph>

          {renderOptions()}
        </Card.Content>
      </LinearGradient>

      <Card.Actions
        style={{
          flexDirection: "row",
          justifyContent: buttonStyle,
        }}
      >
        {renderButtons()}
      </Card.Actions>
    </Card>
  );
};
const styles = StyleSheet.create({
  card: {
    // paddingVertical: 20,
    // paddingHorizontal: 10,
    width: "90%",
    alignSelf: "center",
    borderRadius: 20,
    elevation: 10,
    alignItems: "center",
    justifyContent: "flex-end",
  },
  linearGradient: {
    borderRadius: 20,
    justifyContent: "center",
    paddingVertical: 10,
  },
  question: {
    paddingBottom: 10,
    fontSize: 18,
    fontWeight: "bold",
  },

  optionContainer: {
    width: "75%",
    flexDirection: "row",
    alignSelf: "center",
    alignItems: "center",

    justifyContent: "space-between",
  },
});

PreferenceCard.defaultProps = {
  display: "none",
  question: "",
  restriction: "",
  checkerOptions: [],
  pickerOptions: [],
  firstPage: false,
  lastPage: false,
};
export default PreferenceCard;
