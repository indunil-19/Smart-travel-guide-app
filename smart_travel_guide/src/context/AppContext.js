import React, {createContext, useReducer} from 'react';
import { initialState, reducer as AppReducer } from '../reducer/AppReducer';

export const AppContext =createContext()

export const AppProvider=({children})=>{
    const [state, dispatch]= useReducer(AppReducer, initialState)

    return(
        <AppContext.Provider  value={{state,dispatch}}>
                {children}
        </AppContext.Provider>
    )
}