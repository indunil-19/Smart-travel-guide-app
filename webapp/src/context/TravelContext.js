import React, {createContext, useReducer} from 'react';
import { initialState, Travelreducer } from '../reducer/TravelReducer';

const TravelContext =createContext()

const TravelProvider=({children})=>{
    const [state, dispatch]= useReducer(Travelreducer, initialState)

    return(
        <TravelContext.Provider  value={{state,dispatch}}>
                {children}
        </TravelContext.Provider>
    )
}

export  {TravelContext,  TravelProvider}
 