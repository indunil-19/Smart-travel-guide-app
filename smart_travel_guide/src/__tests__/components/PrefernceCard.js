import React from "react";
import { AppProvider } from "../../context/AppContext";
import { Provider as PaperProvider } from "react-native-paper";
import renderer from "react-test-renderer";
import PreferenceCard from "../../components/PreferenceCard";
test("should match snapshot", () => {
  const tree = renderer
    .create(
      <AppProvider>
        <PaperProvider>
          <PreferenceCard />
        </PaperProvider>
      </AppProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
