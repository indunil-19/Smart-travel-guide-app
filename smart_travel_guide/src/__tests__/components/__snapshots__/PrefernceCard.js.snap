// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "rgba(0, 0, 0, 0.12)",
        "borderRadius": 20,
        "borderWidth": 1,
        "display": "none",
        "elevation": 10,
        "justifyContent": "flex-end",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 9,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 10,
        "width": "90%",
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              4287296741,
              4294967295,
            ],
            "endPoint": undefined,
            "index": 0,
            "locations": undefined,
            "siblings": Array [
              "Card.Actions",
            ],
            "startPoint": undefined,
            "total": 2,
          }
        }
        style={
          Object {
            "borderRadius": 20,
            "justifyContent": "center",
            "paddingVertical": 10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "paddingHorizontal": 16,
              },
              undefined,
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                      "letterSpacing": 0.25,
                      "lineHeight": 20,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 18,
                      "fontWeight": "bold",
                      "paddingBottom": 10,
                    },
                  ],
                ],
              ]
            }
          >
            
          </Text>
          <View>
            <RNCPicker
              items={Array []}
              onChange={[Function]}
              selectedIndex={0}
              style={
                Array [
                  Object {
                    "height": 216,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </ViewManagerAdapter_ExpoLinearGradient>
      <View
        index={1}
        siblings={
          Array [
            "Card.Actions",
          ]
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
              "padding": 8,
            },
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
        total={2}
      >
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "marginLeft": 12,
                    "marginRight": -4,
                  }
                }
              >
                <Text />
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Prev
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0,
              "elevation": 0,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "flexDirection": "row-reverse",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "marginLeft": -4,
                    "marginRight": 12,
                  }
                }
              >
                <Text />
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Next
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
